subsys/fs/nvs/nvs.c

```diff
@@ -0,0 +1,803 @@
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr)));
+	while (1) {
+		rc = nvs_flash_read(fs,
+			_nvs_head_addr_in_flash(fs, &walker),
+			&head,
+			_nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr)));
+		if (rc) {
+			return NVS_ERR_FLASH;
+		}
+		if (head.id == NVS_ID_EMPTY) {
+			return NVS_ERR_NOVAR;
+		}
+		if (head.id == NVS_ID_SECTOR_END) {
+			return NVS_OK;
+		}
+		walker.len = head.len;
+		walker.id = head.id;
+		search.id = walker.id;
+		if (nvs_get_first_entry(fs, &search) != NVS_OK) {
+			/* entry is not found, copy needed - but find the last
+			 * entry first
+			 */
+			walker_last = walker;
+			while (walker_last.id != NVS_ID_SECTOR_END) {
+				rc = nvs_flash_read(fs,
+					_nvs_head_addr_in_flash(fs,
+						&walker_last),
+					&head,
+					_nvs_len_in_flash(fs,
+						sizeof(struct _nvs_data_hdr)));
+				if (rc) {
+					return NVS_ERR_FLASH;
+				}
+				walker_last.len = head.len;
+				walker_last.id = head.id;
+				if (walker_last.id == walker.id) {
+					last_entry = walker_last;
+				}
+				_nvs_addr_advance(fs, &walker_last.data_addr,
+					_nvs_entry_len_in_flash(fs,
+						walker_last.len));
+			}
+			if (last_entry.len == 0) {
+				DBG_NVS("Skipped move of removed entry id %x\n",
+				search.id);
+			} else {
+				rd_addr = _nvs_head_addr_in_flash(fs,
+					&last_entry);
+				len = _nvs_entry_len_in_flash(fs,
+					last_entry.len);
+				while (len > 0) {
+					bytes_to_copy =
+						min(NVS_MOVE_BLOCK_SIZE, len);
+					rc = nvs_flash_read(fs, rd_addr,
+						&buf, bytes_to_copy);
+					if (rc) {
+						return NVS_ERR_FLASH;
+					}
+					rc = nvs_flash_write(fs,
+						fs->write_location,
+						&buf, bytes_to_copy);
+					if (rc) {
+						return NVS_ERR_FLASH;
+					}
+					len -= bytes_to_copy;
+					rd_addr += bytes_to_copy;
+					fs->write_location += bytes_to_copy;
+				}
+				DBG_NVS("Entry with id %x moved to new flash "
+					"sector\n", search.id);
+			}
+		}
+		_nvs_addr_advance(fs, &walker.data_addr,
+			_nvs_entry_len_in_flash(fs, walker.len));
+	}
+}
+
+void nvs_set_start_entry(struct nvs_fs *fs, struct nvs_entry *entry)
+{
+	entry->data_addr = fs->entry_sector * fs->sector_size;
+	_nvs_addr_advance(fs, &entry->data_addr,
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr)) +
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr)));
+}
+
+int nvs_get_first_entry(struct nvs_fs *fs, struct nvs_entry *entry)
+{
+	int rc;
+	struct _nvs_data_hdr head;
+
+	nvs_set_start_entry(fs, entry);
+	while (1) {
+		rc = nvs_flash_read(fs,
+			_nvs_head_addr_in_flash(fs, entry),
+			&head,
+			_nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr)));
+		if (rc) {
+			return NVS_ERR_FLASH;
+		}
+		if (head.id == NVS_ID_EMPTY) {
+			return NVS_ERR_NOVAR;
+		}
+		if (head.id == entry->id) {
+			entry->len = head.len;
+			return NVS_OK;
+		}
+		_nvs_addr_advance(fs,
+			&entry->data_addr,
+			_nvs_entry_len_in_flash(fs, head.len));
+	}
+}
+
+
+int nvs_get_last_entry(struct nvs_fs *fs, struct nvs_entry *entry)
```

Comment:
What do you mean by reverse direction? Walking from the last item to previous? This is not possible as you need the length to jump over items. This could be done if the dog would also contain the length, I have been thinking about that but it would overcomplicate things (and give problems if a write is not complete).

Reply:
> What do you mean by reverse direction? Walking from the last item to previous?

Finally i think it would be optimal only for data which were updated frequently - so does not make any sense to rework this.

thanks for clear this, after rethinking and full-code-review I'vegot similar opinion as you have.

Reply:
Did it make sense? I have seen that my spell checker changed slot into dog :). Sorry for that.

Reply:
I have understood you well. I guessed. I think this function is well implemented, and I agree with:
>  it would overcomplicate things (and give problems if a write is not complete).

So keep the code as is - thanks for explanation.